Upgrade HASH to use Block Protocol 0.2

CiaranMn opened the pull request:
## 🌟 What is the purpose of this PR?

Upgrade HASH to use Block Protocol 0.2

The bulk of the changes are caused by the need to:
1. Updating the function signatures and calls, as per [the new graph specification](https://blockprotocol.org/docs/spec/graph-service-specification) (all the `hooks/blockProtocolFunctions` files)
2. Rewrite the ids and shapes for entities and links.

There are a few more changes detailed below.

This PR depends on a new version of `@blockprotocol/graph` published from https://github.com/blockprotocol/blockprotocol/pull/421 – once both PRs have been reviewed & tested I will cut a further, stable/latest release of `graph` and update the version numbers.

## 🔗 Related links

- [Asana task for updating to 0.2](https://app.asana.com/0/1200211978612931/1202353766855920/f) _(internal)_
- [Task for sending entity types to blocks](https://app.asana.com/0/1200211978612931/1202229645898395/f) _(internal)_

## 🔍 What does this change?

- Rewrite identifiers for entities and links when sending them to blocks: at the moment we need to send `accountId` to our API when identifying entities and links, so this PR introduces rewriting of the `entityId` (and similar) that is sent to blocks so that it contains the accountId (which is then parsed again when the BP functions are called).
- Update shape for entities and types send to blocks: these now have a nested `properties` (or `schema`) object rather than their own properties appearing alongside BP-specified identifiers (`entityId` or `entityTypeId`)
- Starts passing all the `entityTypes` for the entities send to blocks (I've left a [follow-up](https://app.asana.com/0/1202482456346915/1202566631614518/f) to feed through the block entity entity type, which requires a few more changes)
- Updates the HTML block renderer, adds a custom element block renderer
- Update imports for a few utility types from `blockprotocol` to `@blockprotocol/core`
- Fix the entity editor and entity type editor for the new approach
- Remove a couple of deprecated things:
  - `metadataId`: this was an old alias for `entityId` that was only being used in one place, where the page title was being updated.
  - `[blockVariant].displayName`
- comments out some sandbox code, which will be updated for 0.2 as part of [this task](https://app.asana.com/0/1200211978612931/1201901731799594/f)

## 🐾 Next steps

 1. Once this and https://github.com/blockprotocol/blockprotocol/pull/421 have been reviewed & tested I will cut a further, stable/latest release of `graph` and update the version numbers.
 2. Update BP docs to mention HASH as an option for testing blocks
 3. Fix sandboxing in HASH
 4. Eventually fully remove the deprecated `blockprotocol` package from the repo (a few blocks and the WIP new entity editor depend on it)

## 🛡 What tests cover this?

- Existing tests

## ❓ How to test this?

1.  View / create a page with different blocks, check they work
5. Try the entity type editor
6. Try the entity editor (select a type from the sidebar and click 'new [thing]'

CiaranMn commented on packages/hash/api/src/graphql/resolvers/entity/linkGroups.ts:
```diff
@@ -6,15 +6,15 @@ import { Entity, Link } from "../../../model";
 import { DbUnknownEntity } from "../../../db/adapter";
 
-export const DEFAULT_LINK_DEPTH = 2;
+export const DEFAULT_LINK_DEPTH = 1;
```
the `blockGraph` message requires a `depth` parameter. 

I originally made this consistent because I had in my head that `depth` referred to both `linkedEntities` and `linkedAggregations`, but actually `linkedAggregations` doesn't even have a depth parameter. It should, so I've added [a task](https://app.asana.com/0/1201095311341924/1202566735536511/f) to add it.

I'm just going to leave this at 1 so it's consistent with `linkedAggregations` as I expect it'd be easier to reason about.

CiaranMn commented on packages/hash/api/src/graphql/resolvers/entity/linkGroups.ts:
```diff
@@ -35,8 +35,10 @@ export const linkHasSameDestinationAs = (linkA: Link) => (linkB: Link) =>
+
     const existingGroups =
-      linkGroupsObject?.[entity.entityId]?.[entity.entityVersionId];
+      linkGroupsObject?.[`${accountId}*${entityId}`]?.[entityVersionId];
```
We need the `accountId` in this intermediate object to be able to add it to the API output. We need it in the API output to rewrite the `sourceEntityId` to contain the `accountId` before sending it to blocks.

CiaranMn commented on packages/hash/shared/src/queries/entity.queries.ts:
```diff
@@ -170,15 +189,13 @@ export const updateEntity = gql`
       properties: $properties
     ) {
       __typename
-      id
```
We shouldn't be using these ambiguous `id` fields and I want to remove them entirely from the API in a future PR ([task](https://app.asana.com/0/1200211978612931/1202024133311850/f))

nathggns commented:
Looking at this now – but @CiaranMn there's conflicts

nathggns commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolDeleteLink.ts:
```diff
@@ -0,0 +1,71 @@
+} from "../../../graphql/apiTypes.gen";
+import { parseLinkIdentifier } from "../../../lib/entities";
+
+export const useBlockProtocolDeleteLink = (): {
```
These hooks all feel very template-y / copy & paste-y. Do they each individually need to be hooks, or can there be one `useGraphMethods` and each of these functions can be properties on an object returned and can be generated?

CiaranMn replied:
Some of the repetition can be eliminated - I've created [a task](https://app.asana.com/0/1201095311341924/1202584610589613/f) to do so

nathggns commented on packages/hash/frontend/src/lib/entities.ts:
```diff
@@ -1,127 +1,419 @@
+    ApiEntity,
+    "accountId" | "entityId" | "entityTypeId" | "properties"
+  >[],
+): BpEntity[] => records.map(convertApiEntityToBpEntity);
```
`.map` provides extra arguments. You shouldn't pass functions directly to it that aren't specifically designed with that in mind. 

It seems safe, but someone could add a second `number` typed argument to `convertApiEntityToBpEntity` which isn't supposed to be an array index, and TS won't notice the issue. Whereas if you're explicit about what arguments to pass through, it will

This applies to the other functions here too 

```suggestion
): BpEntity[] => records.map(record => convertApiEntityToBpEntity(record));
```

CiaranMn replied:
https://github.com/hashintel/hash/pull/752/commits/e2949ae69542baa0bfddf850b7477ba2f25458d6

teenoh commented on packages/hash/frontend/src/blocks/page/ComponentView.tsx:
```diff
@@ -151,13 +154,13 @@ export class ComponentView implements NodeView<Schema> {
             blockEntityId={entityId}
-            shouldSandbox={!this.editable}
+            blockMetadata={this.componentMetadata}
+            // shouldSandbox={!this.editable}
```
Is commenting this out temporary?

CiaranMn replied:
Done in https://github.com/hashintel/hash/pull/752/commits/359c3d56034d195e626e9eaa7fe0ffdbc3079480

teenoh commented on packages/hash/frontend/src/blocks/page/PageTitle.tsx:
```diff
@@ -59,26 +59,24 @@ export const PageTitle: VoidFunctionComponent<PageTitleProps> = ({
-        data: { title: valueToSave },
+    void updateEntity({
+      data: {
+        entityId: JSON.stringify({ accountId, entityId, entityTypeId: "Page" }),
```
If we haven't, I think we should probably add a comment somewhere that entityId is now a stringified object of accountId, entityId and entityTypeId. I initially assumed we were just combining accountId and entityId

CiaranMn replied:
Thanks for flagging - this is actually left over from an old implementation where I was using `entityTypeId` to identify pages like this. Now it's identified in the call to the hook (passing `true` to indicate it's a page we want to update), so I've removed `entityTypeId` here. It should just be `accountId` and `entityId`

Fixed in https://github.com/hashintel/hash/pull/752/commits/359c3d56034d195e626e9eaa7fe0ffdbc3079480

teenoh commented:
I'm getting this error when I run locally and try to navigate to a new or existing page. 

```
error - TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".json" for /Users/teenoh/Documents/projects/hash/hash/node_modules/@blockprotocol/graph/dist/graph-service.json
```

nathggns reviewed (commented):
Finished code review – will do a functional review now

nathggns commented:
Getting this error when loading a page

```
[dev:frontend] error - SyntaxError: Unexpected identifier
[dev:frontend]     at Loader.moduleStrategy (internal/modules/esm/translators.js:145:18) {
[dev:frontend]   page: '/[account-slug]/[page-slug]'
[dev:frontend] }
```

<img width="910" alt="image" src="https://user-images.githubusercontent.com/719814/178293738-adfe7664-67a3-437b-a746-77a635880187.png">

CiaranMn commented:
> Getting this error when loading a page

@nathggns it works for me locally so I'll get a preview deployment up and see what's going on there - to confirm, you've run `yarn install`? that looks like something I was getting before upgrading `@blockprotocol/graph`

nathggns commented:
@CiaranMn wiped node modules and started from scratch and this happened again. Maybe try that?

CiaranMn commented on packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:
```diff
@@ -1,12 +1,26 @@
+    | "deleteEntity"
+    | "deleteEntityType"
+  >;
+  graphProperties: Required<BlockGraphProperties<Record<string, any>>["graph"]>;
```
Used existing alias for this in https://github.com/hashintel/hash/commit/28a20f86c7fa03177a9591f6da4517518fc534c2

CiaranMn commented on packages/hash/frontend/src/components/RemoteBlock/blockRenderer.tsx:
```diff
@@ -0,0 +1,33 @@
+    return <HtmlLoader html={html} />;
+  }
+  throw new Error(
+    "One of reactElement, customElement or htmlString must be provided.",
```
Thanks, fixed in https://github.com/hashintel/hash/commit/28a20f86c7fa03177a9591f6da4517518fc534c2

CiaranMn commented on packages/hash/frontend/src/pages/[account-slug]/entities/shared/simple-entity-editor.tsx:
```diff
@@ -242,29 +217,28 @@ export const SimpleEntityEditor: VoidFunctionComponent<
+          formData={
+            "entityProperties" in variableProps
+              ? variableProps.entityProperties
+              : {}
+          }
           onSubmit={onSubmit}
-          readonly={readonly}
```
Yes, it was unused. We can add it back in when we implement it properly (which might not be this way)

CiaranMn commented on packages/hash/frontend/src/graphql/queries/entityType.queries.ts:
```diff
@@ -1,6 +1,6 @@
 import { gql } from "@apollo/client";
 
-const entityTypeFieldsFragment = gql`
+export const entityTypeFieldsFragment = gql`
```
Yes, thank you, removed in https://github.com/hashintel/hash/pull/752/commits/426ed11848b1a56a4a2487c2770b93f4b0cdfbd3

CiaranMn commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolCreateEntity.ts:
```diff
@@ -0,0 +1,69 @@
+): {
+  createEntity: EmbedderGraphMessageCallbacks["createEntity"];
+  createEntityLoading: boolean;
+  createEntityError: any;
```
Turns out that all the `Loading/Error` returns were unused, apart from `updateEntityLoading`.

I've removed them all except that and left a `@todo` to consider removing `updateEntityLoading`, since half the hooks didn't have the loading/error returns, and they aren't specified in the BP. 

https://github.com/hashintel/hash/pull/752/commits/18f7e6dad25efbd27ec0a909a9739cbe6f68abfa

CiaranMn commented on packages/hash/frontend/src/components/RemoteBlock/blockRenderer/html.tsx:
```diff
@@ -0,0 +1,34 @@
+      renderHtmlBlock(node, definition, controller.signal).catch((err: any) => {
+        if (err?.name !== "AbortError") {
+          node.innerText = `Error: ${err}`;
+        }
```
Thank you - https://github.com/hashintel/hash/pull/752/commits/6affe4885e685f401069c93d350f5720005c28a2

CiaranMn commented on packages/hash/frontend/src/components/RemoteBlock/blockRenderer/customElement.tsx:
```diff
@@ -0,0 +1,56 @@
+  let existingCustomElement = customElements.get(tagName);
+  if (!existingCustomElement) {
+    try {
+      customElements.define(tagName, elementClass);
```
Thanks for the flag and example - had a go at this in https://github.com/hashintel/hash/pull/752/commits/d22af8b1c1f4e692cc9434e215d43221ce1aaaa1

Only issue is that I had to cast the return of `createComponent` to `any` at the bottom because I was getting "Expression produces a union type that is too complex to represent". It was previously implicitly `any`. I think this is fine for now unless you have any advice on what to do about it.

CiaranMn commented on packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:
```diff
@@ -16,54 +30,99 @@ type RemoteBlockProps = {
+          properties={propsToInject}
+          ReactComponent={
+            (entryPoint === "react" ? blockSource : undefined) as any
+          }
```
Good shout, I think I cast these to sort out later and then never did.

Addressed partially in https://github.com/hashintel/hash/pull/752/commits/d22af8b1c1f4e692cc9434e215d43221ce1aaaa1 in that there's runtime checking that `entryPoint` matches the expected `blockSource` – the compiler isn't narrowing it based on the checking so it could probably be improved

nathggns commented on packages/hash/frontend/src/lib/entities.ts:
```diff
@@ -1,127 +1,419 @@
-  return clone;
+export function parseEntityIdentifier(
+  stringifiedIdentifier: string,
+): ApiEntityIdentifier {
+  let identifierObject: ApiEntityIdentifier;
+  try {
+    identifierObject = JSON.parse(stringifiedIdentifier);
```
So this doesn't look like you're necessarily doing anything wrong. The basic example errors on the playground. Will look into this. 

<img width="913" alt="image" src="https://user-images.githubusercontent.com/719814/178334549-d2605f39-65cf-4324-b5d5-ca74f4311943.png">

<img width="413" alt="image" src="https://user-images.githubusercontent.com/719814/178334555-164dec14-93a7-467c-9bf0-eb98864961f8.png">

nathggns replied:
The explanation / solution is at https://stackoverflow.com/a/70029241/851985 and the Github issue is at https://github.com/microsoft/TypeScript/issues/21732

> This is currently a missing feature in TypeScript. See [microsoft/TypeScript#21732](https://github.com/microsoft/TypeScript/issues/21732) for the relevant feature request. Right now when you use the [in operator as a type guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) like k in result, it only really does any meaningful narrowing if the type of result is a [union](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types) where some members are known to have the key k and others are not. It does not assert the existence of a property with key k. So in your case, where the type of result is just object (narrowed from unknown), nothing meaningful happens.
> 
> The workaround I sometimes use is to write my own [user-defined type guard function](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) which behaves the way I want in to behave. For example:
> 
> function hasKey<K extends string, T extends object>(
>     k: K, o: T
> ): o is T & Record<K, unknown> {
>     return k in o;
> }
> Now, instead of writing k in result, I write hasKey(k, result):
> 
> if (
>     typeof result === "object" &&
>     result !== null &&
>     hasKey("movieName", result) &&
>     typeof result.movieName === "string" // okay
> ) {
>     result.movieName.toUpperCase(); // okay
> }
> This works how you want, and is arguably the closest you can get to type safety here, at least until something is done to address microsoft/TypeScript#21732 natively.

CiaranMn replied:
Very helpful, thank you! Now addressed in https://github.com/hashintel/hash/pull/752/commits/c6e4e2e1c5d6b11ec99b44c4f07e1dcad718e020

CiaranMn commented:
> @CiaranMn wiped node modules and started from scratch and this happened again. Maybe try that?

@nathggns I tried a fresh clone of the repo and it works for me – I _think_ this is related to Node version, as it's also happening in Vercel (which is on 14). what version are you using? Can you try with Node > 18.3.0? It's what I'm using and it's working for me.

CiaranMn commented:
@teenoh @nathggns re. the errors you're getting, I _think_ these are Node version related, as Vercel was giving [an error in building](https://vercel.com/hashintel/hash/Fs6G5fSXVB6P5eyct6uUA15etmGT) until I upgraded the Node version it was using. 

Can you report the Node version you're using that isn't working, and then try `18.3.0`, which is what I'm using? It shouldn't need to be that high (tbh I thought 14 would work), but we can confirm if it's Node version related at least, if that works for you.

teenoh commented on packages/hash/playwright/tests/page-creation.spec.ts:
```diff
@@ -110,7 +110,7 @@ test("user can create page", async ({ page }) => {
   await blockContextMenu
     .locator('[placeholder="Load Block from URL..."]')
-    .fill("https://blockprotocol.org/blocks/@shinypb/emoji-trading-cards");
+    .fill("https://blockprotocol.org/blocks/@hash/code");
```
I think we used a non-default block for this because the context menu initially didn't allow users to re-add an existing block, although that isn't the case now. So we can either

1. Use a non-default block
1. Use a default block, but update the [selector](https://github.com/hashintel/hash/blob/main/packages/hash/playwright/tests/page-creation.spec.ts#L122) in the next line to check for "Re-load block" instead of "Load block"

benwerner01 commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolDeleteLink.ts:
```diff
@@ -0,0 +1,65 @@
+        };
+      }
+
+      return {
+        data: true,
+      };
```
Just curious, is there a reason we aren't using `void` as a return-type for BP methods? If there isn't, perhaps we should consider using `void` as the return type for `delete` methods in a future task, as to my understanding they can only return `true` and never return `false` as they would return errors instead.

benwerner01 commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolUpdateEntity.ts:
```diff
@@ -0,0 +1,73 @@
+import { addIdentifiersToMessageData } from "./shared";
+
+export const useBlockProtocolUpdateEntity = (
+  updateForPage: boolean = false,
```
Perhaps it's worth adding documentation for this parameter to help explain why it's needed (unless we are planning on getting rid of it soon)

CiaranMn replied:
Thanks, done in https://github.com/hashintel/hash/pull/752/commits/60e0ea0fb5ccc616e1a5a588bf8cf4e404afb846

benwerner01 commented on packages/hash/frontend/src/blocks/page/PageTitle.tsx:
```diff
@@ -59,26 +59,24 @@ export const PageTitle: VoidFunctionComponent<PageTitleProps> = ({
-        data: { title: valueToSave },
+    void updateEntity({
+      data: {
+        entityId: JSON.stringify({ accountId, entityId }),
```
Nit:
```suggestion
        entityId: rewriteEntityIdentifier({ accountId, entityId }),
```

benwerner01 commented on packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:
```diff
@@ -16,54 +30,129 @@ type RemoteBlockProps = {
+    graph: graphProperties,
+  };
+
+  // @todo remove this React default when we update all blocks to 0.2
+  const entryPoint = blockMetadata.blockType?.entryPoint ?? "react";
+
+  if (entryPoint === "html" && typeof blockSource !== "string") {
+    throw new Error(
+      `'html' entryPoint expects source to be typeof 'string', but got: ${typeof blockSource}`,
+    );
+  } else if (entryPoint === "custom-element") {
+    if (!(blockSource instanceof HTMLElement)) {
+      throw new Error(
+        `'custom-element' entryPoint expects parsed source to be instanceof 'HTMLElement'`,
+      );
+    }
+    if (typeof blockMetadata.blockType.tagName !== "string") {
+      throw new Error(
+        `Must provide blockType.tagName when entryPoint is 'custom-element'`,
+      );
+    }
+  } else if (entryPoint === "react" && typeof blockSource !== "function") {
+    throw new Error(
+      `'react' entryPoint expects parsed source to be typeof 'function', but got: ${typeof blockSource}`,
     );
+  } else if (!["html", "custom-element", "react"].includes(entryPoint)) {
+    throw new Error(`Invalid entryPoint '${entryPoint}'`);
   }
```
Nit: I could be missing something, but my instinct this logic would probably be better encapsulated within the `BlockRenderer` component? It would probably lead to less casting (as you can directly `return` the JSX elements in each if conditional), and removes the need for the `customElement`/`html`/`ReactComponent` props.

Definitely doesn't block merging, just a thought I'd mention it as it could potentially simplify the logic/prop passing here.

CiaranMn replied:
Thanks @benwerner01 , good suggestion - I've addressed in https://github.com/hashintel/hash/pull/752/commits/5e68957e24b71d5e58c3725778e8321d06682362

There's still a couple of casts in here which could be fixed but I think they're okay for now

benwerner01 replied:
Agreed 👍

benwerner01 commented on packages/hash/frontend/src/components/RemoteBlock/blockRenderer/customElement.tsx:
```diff
@@ -0,0 +1,69 @@
+       * This may break elements that rely on being defined with a specific name.
+       */
+      let i = 0;
+      do {
+        existingCustomElement = customElements.get(tagName);
+        i++;
+      } while (existingCustomElement);
```
**Blocking:** apologies if I'm missing something, but isn't this an infinite loop? I don't think this loop would ever break, because I don't see how incrementing `i` changes whether `existingCustomElement` is defined.

Did you mean to do something like:
```suggestion
      do {
        existingCustomElement = customElements.get(`${tagName}${i}`);
        i++;
      } while (existingCustomElement);
```

benwerner01 commented on packages/hash/frontend/src/lib/entities.ts:
```diff
@@ -1,127 +1,460 @@
+    );
+  }
+
+  if (destinationAccountId.includes("{")) {
+    throw new Error(
+      `destinationAccountId has already been re-written as a stringified object: ${destinationAccountId}`,
+    );
+  }
```
**Blocking:** I could be misunderstanding, but shouldn't we be checking whether `destinationEntityId` is already stringified here? As far as I'm aware we're not stringifying objects into `destinationAccountId` anywhere

benwerner01 commented on packages/hash/frontend/src/lib/entities.ts:
```diff
@@ -1,127 +1,460 @@
+  ApiEntityType,
+  "entityId" | "entityTypeId" | "properties"
+>): BpEntityType => {
+  if (entityTypeId.includes("{")) {
+    throw new Error(
+      `entityTypeId has already been re-written as a stringified object: ${entityTypeId}`,
+    );
+  }
+  return {
```
**Blocking:** similar to my previous comment, why are we checking whether `entityTypeId` is a stringified object if we aren't stringifying it anywhere? Perhaps this should be checking whether `entityId` is stringified?

benwerner01 commented on packages/hash/frontend/src/pages/[account-slug]/entities/shared/simple-entity-editor.tsx:
```diff
@@ -146,52 +140,30 @@ const splitSchema = (
-              ...args.formData,
-            },
-            entityId: entityProps.entityId,
+    if ("entityProperties" in variableProps) {
```
Just for my understanding, why would the check for `entityId` no longer work here?

CiaranMn replied:
It would, either is fine – I probably just rewrote it without thinking. I've updated in https://github.com/hashintel/hash/pull/752/commits/60e0ea0fb5ccc616e1a5a588bf8cf4e404afb846 to consistently use `entityProperties` as the discriminating field to avoid any confusion as to why `entityId` in one place and `entityProperties` in another, but either works.

benwerner01 replied:
Understood, thanks for taking the time think I was overthinking the diff a bit here.

benwerner01 reviewed (changes requested):
Amazing work @CiaranMn 👏

I haven't tested this out locally yet, but wanted to submit my code review sooner as it includes some requested changes which I think are blocking. The majority of my comments are questions/nits which can be ignored and don't require addressing before merging.

nathggns commented:
Looking at this again now

CiaranMn commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolAggregateEntities.ts:
```diff
@@ -1,59 +1,75 @@
+          return {
+            errors: [
+              {
+                code: "INVALID_INPUT",
```
We haven't ruled it out, no, and it was something I thought about when writing these as these errors aren't really appropriate.

I've created a [task](https://app.asana.com/0/1200211978612931/1202591614377625/f) (internal) to add one.

benwerner01 replied:
Makes sense, thanks for creating the task 👍

CiaranMn commented on packages/hash/api/src/graphql/resolvers/entity/linkGroups.ts:
```diff
@@ -6,15 +6,15 @@ import { Entity, Link } from "../../../model";
  * Data structure for collecting the outgoing links
  */
 type LinkGroupsObject = {
-  [sourceEntityId: string]: {
+  [sourceAccountAndEntityId: `${string}*${string}`]: {
```
Haven't had many opportunities to use these yet, but I look for them! ([docs](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html))

nathggns reviewed (dismissed):
Using node 18 definitely works. I wonder if it's connected to namespaced imports somewhere (potentially in a dependency?) – i.e, `import * as path from "node:path"` or whatever? We probably ought to stick to node LTS but also I don't feel *that* strongly about it, but I would like use to put whatever is required in `.nvmrc`. 

Other issues:

- it looks like the image and video blocks aren't working. They hang on uploading. Could you check? 
- changing icon is broken in the callout block
- I get `TypeError: a is not a function` when loading a URL into the embed block
- Syntax highlighting isn't working in the code block – this might be an existing problem

CiaranMn commented on packages/hash/shared/src/queries/entity.queries.ts:
```diff
@@ -1,6 +1,14 @@
+const minimalEntityTypeFieldsFragment = gql`
+  fragment MinimalEntityTypeFields on EntityType {
+    entityId
+    entityTypeId
```
`entityId` is actually sufficient – in the API return, `entityId` is the unique identifier for the entity type.

`entityTypeId` is the type id for _its_ entity type, i.e. the entity type called `EntityType`.

Still not very happy about this - maybe we can take the opportunity of the new Graph API to clear up this confusion.

I've removed `entityTypeId` in https://github.com/hashintel/hash/pull/752/commits/40ba4c9d494f464b16c3c1b95649abba5c212178 - it was originally there when I thought we needed to be converting entity type ids back and forth, but we don't.

CiaranMn commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolAggregateEntities.ts:
```diff
@@ -1,59 +1,75 @@
+  const aggregateEntities: EmbedderGraphMessageCallbacks["aggregateEntities"] =
+    useCallback(
+      async ({ data }) => {
+        if (!data) {
+          return {
+            errors: [
+              {
+                code: "INVALID_INPUT",
+                message: "'data' must be provided for aggregateEntities",
+              },
+            ],
+          };
```
No, it should be treated like this throughout. I've updated `useBlockProtocolUpdateEntity` in https://github.com/hashintel/hash/pull/752/commits/8c0b864bce3bc7cff381ac6e8d9760cefb685102 (and removed the confusing "shared" utility that I only ended up using once)

CiaranMn commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolCreateLink.ts:
```diff
@@ -0,0 +1,79 @@
+export const useBlockProtocolCreateLink = (): {
+  createLink: EmbedderGraphMessageCallbacks["createLink"];
+} => {
+  const [runCreateLinksMutation] = useMutation<
```
Thanks, fixed in https://github.com/hashintel/hash/pull/752/commits/8c0b864bce3bc7cff381ac6e8d9760cefb685102

CiaranMn commented on packages/hash/frontend/src/components/hooks/blockProtocolFunctions/useBlockProtocolAggregateEntities.ts:
```diff
@@ -1,59 +1,75 @@
+        return {
+          data: {
+            operation:
+              returnedOperation as AggregateEntitiesResult<Entity>["operation"],
```
I've created a [task](https://app.asana.com/0/1201095311341924/1202591615077336/f) (internal) – you can't have unions of scalar values in GraphQL, so it'd have to be an enum (or a custom scalar).

CiaranMn commented:
> Using node 18 definitely works. I wonder if it's connected to namespaced imports somewhere (potentially in a dependency?) – i.e, `import * as path from "node:path"` or whatever? We probably ought to stick to node LTS but also I don't feel _that_ strongly about it, but I would like use to put whatever is required in `.nvmrc`.
> 
> Other issues:
> 
> * it looks like the image and video blocks aren't working. They hang on uploading. Could you check?
> * changing icon is broken in the callout block
> * I get `TypeError: a is not a function` when loading a URL into the embed block
> * Syntax highlighting isn't working in the code block – this might be an existing problem

thanks @nathggns 
- I've fixed the file upload in https://github.com/hashintel/hash/pull/752/commits/46bd143bc83392d79aecfc8b8491bcdb41fc5cfd
- Added `.nvmrc` in https://github.com/hashintel/hash/pull/752/commits/5ad0d598d28cf8fc76eaf297baf319521569fddb
- Callout block needs updating to 0.2 - there's [a task to fix it](https://app.asana.com/0/1200211978612931/1202418951550823/f) as part of updating it to use the hook service, and I'd like to leave it for then
- Embed block was relying on a weird `fetchEmbedCode` function that we've stopped specifying in the BP - there's a task to address this [here](https://app.asana.com/0/1200211978612931/1202509819279267/f)
- Happy to filter these two broken blocks out of the block suggester and mock data in a follow up if we think best
- Syntax highlighting is working for me in the code block
<img width="657" alt="Screenshot 2022-07-12 at 16 38 34" src="https://user-images.githubusercontent.com/37743469/178542402-5f0f1903-32bb-41a7-b1a6-60a93f108bef.png">

benwerner01 reviewed (dismissed):
Thank you for taking the time to respond/address all my comments, I've tested it locally and couldn't spot any issues 👍

nonparibus commented:
@CiaranMn Is there a reason we've added `.nvmrc` at the top-level instead of in the `packages/hash` subdirectory? cc @kachkaev

teenoh commented on packages/hash/playwright/tests/page-creation.spec.ts:
```diff
@@ -119,10 +119,10 @@ test("user can create page", async ({ page }) => {
 
   await expect(
-    blockContextMenu.locator('[placeholder="Load Block from URL..."]'),
+    blockContextMenu.locator('[placeholder="Re-load block"]'),
```
This should be "Load Block from URL..." since it's the input's placeholder and not the button 


```suggestion
    blockContextMenu.locator('[placeholder="Load Block from URL..."]'),
```

nonparibus replied:
Block should not be capitalized where it appears in the middle of a sentence, outside of the brand name/proper noun "Block Protocol".

kachkaev commented:
Good point about `.nvmrc` @nonparibus! I’m not against having this file even though not everyone uses `nvm`. Perhaps, we could also specify `package.json` → `engines` and make the `.nvmrc` version less precise, like in https://github.com/blockprotocol/blockprotocol/pull/330.

CiaranMn commented:
@nonparibus @kachkaev I have moved `.nvmrc` to the `frontend` folder and updated it, and added `engines`. 

I don't know what versions don't work but the latest 16 definitely does.

Worth noting that neither `.nvmrc` nor `engines` in package.json will do anything unless the user has configured them to do, but they provide a hint at least.

teenoh reviewed (approved):
Thanks for addressing my feedback @CiaranMn! Tested and things are working well locally

kachkaev commented:
Not sure if `.nvmrc` and `package.json` → `engines` are ever read from a yarn workspace. I’ve only seen them in project root folder before, so they are probably a noop. Maybe we can just remove those lines from this PR as they are out of scope in a way.

nonparibus commented:
> Not sure if `.nvmrc` and `package.json` → `engines` are ever read from a yarn workspace. I’ve only seen them in project root folder before, so they are probably a noop. Maybe we can just remove those lines from this PR as they are out of scope in a way.

I'd advocate we keep it in. [Some](https://github.com/renovatebot/renovate/pull/6107) tooling supports nested `nvmrc` files. `nvm` themselves mention that "any parent directory" can be used in lieu of a repo root (in their [README](https://github.com/nvm-sh/nvm#nvmrc=)).

teenoh commented:
> Block should not be capitalized. @teenoh @CiaranMn See: [#752 (comment)](https://github.com/hashintel/hash/pull/752#discussion_r919854050)

Addressed in https://github.com/hashintel/hash/pull/796
